app/components/inputs/hooks/useUploadImage.ts

```diff
@@ -0,0 +1,143 @@
+import { ImageListType, ImageType } from "react-images-uploading";
+import { useState } from "react";
+import axios from "axios";
+import { toast } from "react-hot-toast";
+import { PreviewImageWithUrl } from "@/app/types";
+
+interface Props {
+  value: PreviewImageWithUrl[];
+  onChange: (value: PreviewImageWithUrl[]) => void;
+}
+
+export const useUploadImage = ({ value: images, onChange }: Props) => {
+  const [previewImages, setPreviewImages] = useState(
+    images.map((image) => {
+      const { url, ...rest } = image;
+      return { ...rest, dataURL: url as string } as ImageType;
```

Comment:
The two urls can be a little bit confusing, but I need them both to make it work. When the image is being uploaded to Cloudinary, I'm setting the image source to `dataURL` temporarily to be able to still display the image, but I also need a field to store the unresolved remote url, that's what the field `url` is for. Once the remote url in Cloudinary is ready, I'm changing `dataURL` to the remote url so what you see on the screen is exactly what the final data looks like. By doing this, I'm also able to know when the upload is done in Cloudinary by checking if `dataURL` is local or remote. That's the general logic for the whole upload thing.

In this case specifically, `dataURL` has to be the value of `url` in `image` so that the image source is remote, indicating that the image is done uploading to Cloudinary